rfhub2/cli/test_cases/test_cases_extractor.py

```diff
@@ -0,0 +1,121 @@
+from dataclasses import dataclass
+from pathlib import Path
+from robot.running.model import TestSuite
+from robot.testdoc import TestSuiteFactory
+from typing import List, Tuple
+
+
+@dataclass
+class Keyword:
+    name: str
+
+
+@dataclass
+class Test:
+    doc: str
+    id: str
+    longname: str
+    name: str
+    parent: str
+    source: str
+    tags: List[str]
+    template: str
+    timeout: str
+    keywords: List[Keyword]
+    setup: List[Keyword]
+    teardown: List[Keyword]
+
+
+@dataclass
+class Suite:
+    doc: str
+    id: str
+    longname: str
+    name: str
+    parent: str
+    source: str
+    test_count: int
+    tests: List[Test]
+    setup: List[Keyword]
+    teardown: List[Keyword]
+
+
+class TestCasesExtractor:
+    def __init__(self, paths: Tuple[Path, ...]) -> None:
+        self.paths = paths
+
+    def create_testdoc_from_paths(self):
+        """
+        Creates TestDoc objects based on robot.testdoc.TestSuiteFactory method.
+        """
+        return [suite for path in self.paths for suite in self.create_testdoc_from_path(path)]
+
+    def create_testdoc_from_path(self, path: Path):
+        """
+        Creates TestDoc object based on robot.testdoc.TestSuiteFactory method.
+        """
+        testdoc = TestSuiteFactory(str(path))
+        return [self.serialize_testdoc(testdoc)] + self._traverse_suites(testdoc)

```

Comment:
It looks like this first serialize_testdoc call could be part of _traverse_suites method if it would first serialize_testdoc of top level suite and then for each child suite call traverse_suites and combine results

Reply:
Wouldn't that end in duplicating suite objects? If top level serialization would be added at the beginning, it would treat sub-suite as top level suite for every method call. Or am I wrong? My force is not that strong when recurrence calls.

Reply:
I figured out what You meant. It works! :)